File: 0056-merge-intervals/0056-merge-intervals.py
from typing import List

class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        if not intervals:
            return []

        # Sort intervals by the start time
        intervals.sort()
        result = []
        
        # Start with the first interval as the current newInterval
        newInterval = intervals[0]
        result.append(newInterval)
        
        for i in range(1, len(intervals)):
            currentStart, currentEnd = intervals[i]
            
            if currentStart <= newInterval[1]:  # Check for overlap
                # Merge intervals by updating the end of newInterval
                newInterval[1] = max(newInterval[1], currentEnd)
            else:
                # No overlap, add newInterval to result and start a new interval
                newInterval = intervals[i]
                result.append(newInterval)
        
        return result

File: 0056-merge-intervals/0056-merge-intervals.java
import java.util.*;

class Solution {
    public int[][] merge(int[][] intervals) {

        if (intervals.length <= 1) {
            return intervals;
        }

        // 1️⃣ Sort intervals by start time
        Arrays.sort(intervals, Comparator.comparingInt(i -> i[0]));

        List<int[]> result = new ArrayList<>();
        int[] newInterval = intervals[0]; // start with the first interval
        result.add(newInterval);          // ✅ add it immediately

        // 2️⃣ Merge or add
        for (int[] interval : intervals) {
            if (interval[0] <= newInterval[1]) {   // overlapping
                newInterval[1] = Math.max(newInterval[1], interval[1]);
            } else {
                newInterval = interval;            // move to next non-overlapping
                result.add(newInterval);
            }
        }

        // 3️⃣ Convert list to array
        return result.toArray(new int[result.size()][]);
    }
}
